baseplate/context/redis.py

```diff
@@ -4,55 +4,83 @@
-
-    .. note::
-
-        Pipelines, transactions, locks, and pubsub are currently unsupported.
+    This factory will attach a
+    :py:class:`~baseplate.context.redis.MonitoredRedisConnection` to an
+    attribute on the :term:`context object`. When redis commands are executed
+    via this connection object, they will use connections from the provided
+    :py:class:`redis.ConnectionPool` and automatically record diagnostic
+    information.
 
     :param redis.ConnectionPool connection_pool: A connection pool.
 
+    :returns: :py:class:`~baseplate.context.redis.MonitoredRedisConnection`
+
     """
     def __init__(self, connection_pool):
         self.connection_pool = connection_pool
 
     def make_object_for_context(self, name, root_span):
-        return RedisConnectionAdapter(name, root_span, self.connection_pool)
+        return MonitoredRedisConnection(name, root_span, self.connection_pool)
+
+
+class MonitoredRedisConnection(redis.StrictRedis):
+    """Redis connection that collects diagnostic information.
 
+    This connection acts like :py:class:`redis.StrictRedis` except that all
+    operations are automatically wrapped with diagnostic collection.
+
+    The interface is the same as that class except for the
+    :py:meth:`~baseplate.context.redis.MonitoredRedisConnection.pipeline`
+    method.
+
+    .. note:: Locks and pubsub are currently unsupported.
+
+    """
 
-class RedisConnectionAdapter(redis.StrictRedis):
     def __init__(self, context_name, root_span, connection_pool):
         self.context_name = context_name
         self.root_span = root_span
 
-        super(RedisConnectionAdapter, self).__init__(
+        super(MonitoredRedisConnection, self).__init__(
             connection_pool=connection_pool)
 
     def execute_command(self, command, *args, **kwargs):
         trace_name = "{}.{}".format(self.context_name, command)
 
         with self.root_span.make_child(trace_name):
-            return super(RedisConnectionAdapter, self).execute_command(
+            return super(MonitoredRedisConnection, self).execute_command(
                 command, *args, **kwargs)
 
-    # these commands are not yet implemented because i haven't taken the time
-    # to look at if they go through `execute_command` like the basic commands
-    # (e.g. "get" and "set") do. if these commands end up needed, it should
-    # just be some elbow grease to get them going.
-    def pipeline(self, *args, **kwargs):
-        raise NotImplementedError
+    def pipeline(self, name, transaction=True, shard_hint=None):
+        """Create a pipeline.
+
+        This returns an object on which you can call the standard redis
+        commands. Execution will be deferred until ``execute`` is called. This
+        is useful for saving round trips.
+
+        :param str name: The name to attach to diagnostics for this pipeline.
+        :param bool transaction: Whether or not the commands in the pipeline
```

Comment:
So will this work currently? I see `NotImplementedError`, but is that only if you're using `pipeline` directly?